File: 4119-minimum-distance-between-three-equal-elements-ii/minimum-distance-between-three-equal-elements-ii.cpp
class Solution {
public:
    int minimumDistance(vector<int>& nums) {
        // if i use sliding window 

        unordered_map<int,int> mp;
        int n = nums.size();

        int j = 0;
        int i = 0;

        int ans = INT_MAX;
        while(j<n){
            int num = nums[j];
            // cout<<"num: "<<num<<endl;
            mp[num]++;
            // cout<<mp[num]<<endl;
            if(mp[num]==3){
                ans= min(ans, (j-i)*2);
                
                while(nums[i]!=num){
                    mp[nums[i]]--;
                    i++;
                }
                ans= min(ans, (j-i)*2);
                mp[nums[i]]--;
                i++;

            }
            j++;
        }

        return ans== INT_MAX?-1:ans;

    }
};

File: graphs-and-paths/4119-minimum-distance-between-three-equal-elements-ii/minimum-distance-between-three-equal-elements-ii.java
class Solution {
    public int minimumDistance(int[] nums) {
        int n = nums.length;

        int i = 0;
        int j = 0;

        int ans = Integer.MAX_VALUE;

        HashMap<Integer, Integer> mp = new HashMap<>();

        while(j<n){

            int num = nums[j];

            mp.put(num,mp.getOrDefault(num,0)+1);

            if(mp.get(num)==3){
                while(nums[i]!=num){
                    mp.put(nums[i], mp.get(nums[i]) - 1);
                     i++;

                }
                ans = Math.min(ans, (j - i) * 2); 
                mp.put(nums[i], mp.get(nums[i]) - 1);
                i++;
            }
            j++;

        }
         return ans == Integer.MAX_VALUE ? -1 : ans;
    }
}